File: frontend/src/types/certifications.ts
/**
 * Certification & Compliance Hub Types
 *
 * Types for managing producer certifications (GOST, organic, halal, kosher, eco-labels)
 */

// =============================================================================
// Enums
// =============================================================================

export type CertificationCategory =
  | 'quality_standard'  // GOST, ISO
  | 'organic'           // Organic certifications
  | 'religious'         // Halal, Kosher
  | 'eco_label'         // Eco-friendly labels
  | 'safety'            // Sanitary, safety
  | 'origin'            // PDO, PGI, geographic
  | 'other'

export type VerificationStatus =
  | 'pending'        // Awaiting review
  | 'verified'       // Confirmed by chestno.ru
  | 'rejected'       // Failed verification
  | 'expired'        // Past expiry date
  | 'revoked'        // Revoked by issuing body
  | 'auto_verified'  // Verified via API

export type VerificationRequestType =
  | 'authenticity_check'
  | 'dispute'
  | 'renewal_inquiry'

export type VerificationRequestStatus =
  | 'open'
  | 'investigating'
  | 'resolved'
  | 'dismissed'

// =============================================================================
// Certification Types (Reference Data)
// =============================================================================

export interface CertificationType {
  id: string
  code: string
  name_ru: string
  name_en: string
  description_ru?: string
  description_en?: string
  category: CertificationCategory
  issuing_body_name_ru?: string
  issuing_body_name_en?: string
  issuing_body_website?: string
  issuing_body_country: string
  logo_url?: string
  badge_color: string
  requires_document: boolean
  auto_verify_enabled: boolean
  is_active: boolean
  display_order: number
}

export interface CertificationTypePublic {
  id: string
  code: string
  name_ru: string
  name_en: string
  category: CertificationCategory
  logo_url?: string
  badge_color: string
}

// =============================================================================
// Producer Certifications
// =============================================================================

export interface ProducerCertificationCreate {
  certification_type_id: string
  certificate_number?: string
  issued_by?: string
  issued_date?: string  // ISO date
  expiry_date?: string  // ISO date
  scope_description?: string
  product_ids?: string[]
  is_public?: boolean
  display_on_products?: boolean
}

export interface ProducerCertificationUpdate {
  certificate_number?: string
  issued_by?: string
  issued_date?: string
  expiry_date?: string
  scope_description?: string
  product_ids?: string[]
  is_public?: boolean
  display_on_products?: boolean
}

export interface ProducerCertification {
  id: string
  organization_id: string
  certification_type_id: string
  certification_type?: CertificationTypePublic

  // Certificate details
  certificate_number?: string
  issued_by?: string
  issued_date?: string
  expiry_date?: string
  scope_description?: string
  product_ids?: string[]

  // Document
  document_url?: string
  document_original_name?: string
  document_uploaded_at?: string

  // Verification
  verification_status: VerificationStatus
  verification_notes?: string
  verified_at?: string

  // Visibility
  is_public: boolean
  display_on_products: boolean

  // Computed
  is_valid: boolean
  days_until_expiry?: number

  created_at: string
  updated_at: string
}

export interface ProducerCertificationPublic {
  id: string
  certification_type: CertificationTypePublic
  certificate_number?: string
  issued_by?: string
  issued_date?: string
  expiry_date?: string
  scope_description?: string
  verification_status: VerificationStatus
  is_valid: boolean
}

// =============================================================================
// Document Management
// =============================================================================

export interface DocumentUploadResponse {
  document_url: string
  original_name: string
  uploaded_at: string
  file_size_bytes: number
}

export interface DocumentVerificationResult {
  is_valid: boolean
  confidence_score: number
  extracted_data: Record<string, unknown>
  warnings: string[]
  errors: string[]
}

// =============================================================================
// Verification
// =============================================================================

export interface VerifyExternalResponse {
  is_verified: boolean
  verification_source: string
  external_id?: string
  verification_date: string
  valid_until?: string
  details: Record<string, unknown>
}

export interface VerificationLogEntry {
  id: string
  certification_id: string
  action: string
  previous_status?: string
  new_status: string
  notes?: string
  performed_by?: string
  performed_at: string
}

// =============================================================================
// Expiry Alerts
// =============================================================================

export interface ExpiryAlert {
  id: string
  certification_id: string
  organization_id: string
  certification_type_name: string
  certificate_number?: string
  expiry_date: string
  alert_days_before: number
  scheduled_at: string
  sent_at?: string
  acknowledged_at?: string
}

export interface ExpiryAlertSettings {
  alert_days: number[]
  email_enabled: boolean
  push_enabled: boolean
}

// =============================================================================
// Consumer Verification Portal
// =============================================================================

export interface ConsumerVerificationRequest {
  certification_id: string
  request_type: VerificationRequestType
  reason?: string
}

export interface CertificationSearchFilters {
  certification_types?: string[]
  categories?: CertificationCategory[]
  verified_only?: boolean
  include_expired?: boolean
}

export interface CertificationSearchResult {
  product_id: string
  product_name: string
  organization_id: string
  organization_name: string
  certifications: ProducerCertificationPublic[]
}

export interface CertificationSearchResponse {
  results: CertificationSearchResult[]
  total: number
  page: number
  page_size: number
  has_more: boolean
}

// =============================================================================
// Organization Summary
// =============================================================================

export interface OrganizationCertificationSummary {
  organization_id: string
  total_certifications: number
  verified_count: number
  pending_count: number
  expiring_soon_count: number
  expired_count: number
  certifications_by_category: Record<string, number>
  certifications: ProducerCertificationPublic[]
}

// =============================================================================
// Admin Dashboard
// =============================================================================

export interface CertificationAdminStats {
  total_certifications: number
  pending_verification: number
  verified_today: number
  expiring_this_week: number
  open_disputes: number
  by_status: Record<string, number>
  by_category: Record<string, number>
  recent_submissions: ProducerCertification[]
}

export interface PendingVerificationItem {
  certification: ProducerCertification
  organization_name: string
  submitted_at: string
  days_pending: number
  document_available: boolean
}

// =============================================================================
// Display Configuration
// =============================================================================

export const VERIFICATION_STATUS_CONFIG: Record<VerificationStatus, {
  label_ru: string
  label_en: string
  color: string
  bgColor: string
  icon: string
}> = {
  pending: {
    label_ru: 'На проверке',
    label_en: 'Pending',
    color: 'text-yellow-700',
    bgColor: 'bg-yellow-100',
    icon: 'clock',
  },
  verified: {
    label_ru: 'Подтверждён',
    label_en: 'Verified',
    color: 'text-green-700',
    bgColor: 'bg-green-100',
    icon: 'check-circle',
  },
  auto_verified: {
    label_ru: 'Автоподтверждён',
    label_en: 'Auto-verified',
    color: 'text-green-700',
    bgColor: 'bg-green-100',
    icon: 'check-badge',
  },
  rejected: {
    label_ru: 'Отклонён',
    label_en: 'Rejected',
    color: 'text-red-700',
    bgColor: 'bg-red-100',
    icon: 'x-circle',
  },
  expired: {
    label_ru: 'Истёк',
    label_en: 'Expired',
    color: 'text-gray-700',
    bgColor: 'bg-gray-100',
    icon: 'exclamation-circle',
  },
  revoked: {
    label_ru: 'Отозван',
    label_en: 'Revoked',
    color: 'text-red-700',
    bgColor: 'bg-red-100',
    icon: 'ban',
  },
}

export const CATEGORY_CONFIG: Record<CertificationCategory, {
  label_ru: string
  label_en: string
  icon: string
  color: string
}> = {
  quality_standard: {
    label_ru: 'Стандарт качества',
    label_en: 'Quality Standard',
    icon: 'badge-check',
    color: '#1E40AF',
  },
  organic: {
    label_ru: 'Органик',
    label_en: 'Organic',
    icon: 'leaf',
    color: '#22C55E',
  },
  religious: {
    label_ru: 'Религиозный',
    label_en: 'Religious',
    icon: 'star',
    color: '#10B981',
  },
  eco_label: {
    label_ru: 'Эко-маркировка',
    label_en: 'Eco Label',
    icon: 'globe',
    color: '#84CC16',
  },
  safety: {
    label_ru: 'Безопасность',
    label_en: 'Safety',
    icon: 'shield-check',
    color: '#EF4444',
  },
  origin: {
    label_ru: 'Происхождение',
    label_en: 'Origin',
    icon: 'map-pin',
    color: '#8B5CF6',
  },
  other: {
    label_ru: 'Другое',
    label_en: 'Other',
    icon: 'document-text',
    color: '#6B7280',
  },
}

// Common Russian certifications with their codes
export const COMMON_CERTIFICATIONS = {
  // Quality Standards
  GOST_R: 'gost_r',
  GOST_ISO: 'gost_iso',
  ISO_9001: 'iso_9001',
  ISO_22000: 'iso_22000',
  ROSKACHESTVO: 'roskachestvo',

  // Organic
  ORGANIC_RU: 'organic_ru',
  ORGANIC_EU: 'organic_eu',
  USDA_ORGANIC: 'usda_organic',
  ECOCERT: 'ecocert',

  // Religious
  HALAL_RU: 'halal_ru',
  HALAL_ISWA: 'halal_iswa',
  KOSHER_RU: 'kosher_ru',
  KOSHER_OU: 'kosher_ou',

  // Eco Labels
  LEAF_OF_LIFE: 'leaf_of_life',
  ECO_PRODUCT_RU: 'eco_product_ru',
  FSC: 'fsc',
  RAINFOREST_ALLIANCE: 'rainforest_alliance',

  // Safety
  SANITARY_RU: 'sanitary_ru',
  DECLARATION_CONFORMITY: 'declaration_conformity',
  FIRE_SAFETY: 'fire_safety',

  // Origin
  PDO_RU: 'pdo_ru',
  PGI_RU: 'pgi_ru',
} as const

File: backend/app/schemas/certifications.py
"""
Certification & Compliance Hub Schemas.

Defines Pydantic models for the certification management system.
Supports GOST, organic, halal, kosher, eco-labels, and other certifications.
"""
from datetime import date, datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field, HttpUrl


class CertificationCategory(str, Enum):
    """Categories of certifications."""
    QUALITY_STANDARD = "quality_standard"   # GOST, ISO, etc.
    ORGANIC = "organic"                     # Organic certifications
    RELIGIOUS = "religious"                 # Halal, Kosher
    ECO_LABEL = "eco_label"                 # Eco-friendly labels
    SAFETY = "safety"                       # Sanitary, safety certs
    ORIGIN = "origin"                       # PDO, PGI, geographic
    OTHER = "other"


class VerificationStatus(str, Enum):
    """Verification status of a certification."""
    PENDING = "pending"             # Awaiting review
    VERIFIED = "verified"           # Confirmed by chestno.ru
    REJECTED = "rejected"           # Failed verification
    EXPIRED = "expired"             # Past expiry date
    REVOKED = "revoked"             # Revoked by issuing body
    AUTO_VERIFIED = "auto_verified" # Verified via API


class VerificationRequestType(str, Enum):
    """Types of consumer verification requests."""
    AUTHENTICITY_CHECK = "authenticity_check"
    DISPUTE = "dispute"
    RENEWAL_INQUIRY = "renewal_inquiry"


class VerificationRequestStatus(str, Enum):
    """Status of verification requests."""
    OPEN = "open"
    INVESTIGATING = "investigating"
    RESOLVED = "resolved"
    DISMISSED = "dismissed"


# =============================================================================
# Certification Types (Reference Data)
# =============================================================================

class CertificationTypeBase(BaseModel):
    """Base certification type info."""
    code: str
    name_ru: str
    name_en: str
    description_ru: Optional[str] = None
    description_en: Optional[str] = None
    category: CertificationCategory
    issuing_body_name_ru: Optional[str] = None
    issuing_body_name_en: Optional[str] = None
    issuing_body_website: Optional[str] = None
    issuing_body_country: str = "RU"
    logo_url: Optional[str] = None
    badge_color: str = "#4F46E5"


class CertificationType(CertificationTypeBase):
    """Full certification type with metadata."""
    id: str
    requires_document: bool = True
    auto_verify_enabled: bool = False
    is_active: bool = True
    display_order: int = 100
    created_at: datetime
    updated_at: datetime


class CertificationTypePublic(BaseModel):
    """Public view of certification type."""
    id: str
    code: str
    name_ru: str
    name_en: str
    category: CertificationCategory
    logo_url: Optional[str] = None
    badge_color: str


# =============================================================================
# Producer Certifications
# =============================================================================

class ProducerCertificationCreate(BaseModel):
    """Request to add a new certification."""
    certification_type_id: str
    certificate_number: Optional[str] = None
    issued_by: Optional[str] = None
    issued_date: Optional[date] = None
    expiry_date: Optional[date] = None
    scope_description: Optional[str] = None
    product_ids: Optional[list[str]] = None
    is_public: bool = True
    display_on_products: bool = True


class ProducerCertificationUpdate(BaseModel):
    """Request to update a certification."""
    certificate_number: Optional[str] = None
    issued_by: Optional[str] = None
    issued_date: Optional[date] = None
    expiry_date: Optional[date] = None
    scope_description: Optional[str] = None
    product_ids: Optional[list[str]] = None
    is_public: Optional[bool] = None
    display_on_products: Optional[bool] = None


class ProducerCertification(BaseModel):
    """Full producer certification record."""
    id: str
    organization_id: str
    certification_type_id: str
    certification_type: Optional[CertificationTypePublic] = None

    # Certificate details
    certificate_number: Optional[str] = None
    issued_by: Optional[str] = None
    issued_date: Optional[date] = None
    expiry_date: Optional[date] = None
    scope_description: Optional[str] = None
    product_ids: Optional[list[str]] = None

    # Document
    document_url: Optional[str] = None
    document_original_name: Optional[str] = None
    document_uploaded_at: Optional[datetime] = None

    # Verification
    verification_status: VerificationStatus = VerificationStatus.PENDING
    verification_notes: Optional[str] = None
    verified_at: Optional[datetime] = None

    # Visibility
    is_public: bool = True
    display_on_products: bool = True

    # Computed
    is_valid: bool = False  # True if verified and not expired
    days_until_expiry: Optional[int] = None

    created_at: datetime
    updated_at: datetime


class ProducerCertificationPublic(BaseModel):
    """Public view of a certification (for consumers)."""
    id: str
    certification_type: CertificationTypePublic
    certificate_number: Optional[str] = None
    issued_by: Optional[str] = None
    issued_date: Optional[date] = None
    expiry_date: Optional[date] = None
    scope_description: Optional[str] = None
    verification_status: VerificationStatus
    is_valid: bool


# =============================================================================
# Document Upload
# =============================================================================

class DocumentUploadResponse(BaseModel):
    """Response after uploading a certification document."""
    document_url: str
    original_name: str
    uploaded_at: datetime
    file_size_bytes: int


class DocumentVerificationResult(BaseModel):
    """Result of automated document verification."""
    is_valid: bool
    confidence_score: float  # 0.0 to 1.0
    extracted_data: dict = Field(default_factory=dict)
    warnings: list[str] = Field(default_factory=list)
    errors: list[str] = Field(default_factory=list)


# =============================================================================
# Verification
# =============================================================================

class VerifyExternalRequest(BaseModel):
    """Request to verify against external certification body."""
    certification_id: str
    certificate_number: str


class VerifyExternalResponse(BaseModel):
    """Response from external verification."""
    is_verified: bool
    verification_source: str  # e.g., "rosstandart_api"
    external_id: Optional[str] = None
    verification_date: datetime
    valid_until: Optional[date] = None
    details: dict = Field(default_factory=dict)
    raw_response: Optional[dict] = None


class ManualVerificationRequest(BaseModel):
    """Admin request to manually verify/reject a certification."""
    certification_id: str
    action: str  # 'verify', 'reject', 'revoke'
    notes: Optional[str] = None


class VerificationLogEntry(BaseModel):
    """Single verification log entry."""
    id: str
    certification_id: str
    action: str
    previous_status: Optional[str] = None
    new_status: str
    notes: Optional[str] = None
    performed_by: Optional[str] = None
    performed_at: datetime


# =============================================================================
# Expiry Alerts
# =============================================================================

class ExpiryAlert(BaseModel):
    """Certification expiry alert."""
    id: str
    certification_id: str
    organization_id: str
    certification_type_name: str
    certificate_number: Optional[str] = None
    expiry_date: date
    alert_days_before: int
    scheduled_at: datetime
    sent_at: Optional[datetime] = None
    acknowledged_at: Optional[datetime] = None


class ExpiryAlertSettings(BaseModel):
    """Settings for expiry alerts."""
    alert_days: list[int] = [90, 60, 30, 14, 7, 1]
    email_enabled: bool = True
    push_enabled: bool = True


class AcknowledgeAlertRequest(BaseModel):
    """Request to acknowledge an expiry alert."""
    alert_id: str


# =============================================================================
# Consumer Verification Portal
# =============================================================================

class ConsumerVerificationRequest(BaseModel):
    """Consumer request to verify a certification."""
    certification_id: str
    request_type: VerificationRequestType
    reason: Optional[str] = None


class ConsumerVerificationResponse(BaseModel):
    """Response to consumer verification request."""
    request_id: str
    status: VerificationRequestStatus
    certification: ProducerCertificationPublic
    message: str


class CertificationSearchFilters(BaseModel):
    """Filters for searching products by certification."""
    certification_types: Optional[list[str]] = None  # cert type codes
    categories: Optional[list[CertificationCategory]] = None
    verified_only: bool = True
    include_expired: bool = False


class CertificationSearchResult(BaseModel):
    """Single search result."""
    product_id: str
    product_name: str
    organization_id: str
    organization_name: str
    certifications: list[ProducerCertificationPublic]


class CertificationSearchResponse(BaseModel):
    """Paginated search results."""
    results: list[CertificationSearchResult]
    total: int
    page: int
    page_size: int
    has_more: bool


# =============================================================================
# Organization Certification Summary
# =============================================================================

class OrganizationCertificationSummary(BaseModel):
    """Summary of organization's certifications."""
    organization_id: str
    total_certifications: int
    verified_count: int
    pending_count: int
    expiring_soon_count: int  # Within 30 days
    expired_count: int
    certifications_by_category: dict[str, int]
    certifications: list[ProducerCertificationPublic]


# =============================================================================
# Admin Dashboard
# =============================================================================

class CertificationAdminStats(BaseModel):
    """Admin dashboard statistics."""
    total_certifications: int
    pending_verification: int
    verified_today: int
    expiring_this_week: int
    open_disputes: int
    by_status: dict[str, int]
    by_category: dict[str, int]
    recent_submissions: list[ProducerCertification]


class PendingVerificationItem(BaseModel):
    """Item in pending verification queue."""
    certification: ProducerCertification
    organization_name: str
    submitted_at: datetime
    days_pending: int
    document_available: bool


class PendingVerificationQueue(BaseModel):
    """Queue of certifications awaiting verification."""
    items: list[PendingVerificationItem]
    total: int
    page: int
    page_size: int


# =============================================================================
# Status Display Configuration
# =============================================================================

VERIFICATION_STATUS_CONFIG = {
    VerificationStatus.PENDING: {
        "label_ru": "На проверке",
        "label_en": "Pending",
        "color": "text-yellow-700",
        "bg_color": "bg-yellow-100",
        "icon": "clock",
    },
    VerificationStatus.VERIFIED: {
        "label_ru": "Подтверждён",
        "label_en": "Verified",
        "color": "text-green-700",
        "bg_color": "bg-green-100",
        "icon": "check-circle",
    },
    VerificationStatus.AUTO_VERIFIED: {
        "label_ru": "Автоподтверждён",
        "label_en": "Auto-verified",
        "color": "text-green-700",
        "bg_color": "bg-green-100",
        "icon": "check-badge",
    },
    VerificationStatus.REJECTED: {
        "label_ru": "Отклонён",
        "label_en": "Rejected",
        "color": "text-red-700",
        "bg_color": "bg-red-100",
        "icon": "x-circle",
    },
    VerificationStatus.EXPIRED: {
        "label_ru": "Истёк",
        "label_en": "Expired",
        "color": "text-gray-700",
        "bg_color": "bg-gray-100",
        "icon": "exclamation-circle",
    },
    VerificationStatus.REVOKED: {
        "label_ru": "Отозван",
        "label_en": "Revoked",
        "color": "text-red-700",
        "bg_color": "bg-red-100",
        "icon": "ban",
    },
}

CATEGORY_DISPLAY_CONFIG = {
    CertificationCategory.QUALITY_STANDARD: {
        "label_ru": "Стандарт качества",
        "label_en": "Quality Standard",
        "icon": "badge-check",
    },
    CertificationCategory.ORGANIC: {
        "label_ru": "Органик",
        "label_en": "Organic",
        "icon": "leaf",
    },
    CertificationCategory.RELIGIOUS: {
        "label_ru": "Религиозный",
        "label_en": "Religious",
        "icon": "star",
    },
    CertificationCategory.ECO_LABEL: {
        "label_ru": "Эко-маркировка",
        "label_en": "Eco Label",
        "icon": "globe",
    },
    CertificationCategory.SAFETY: {
        "label_ru": "Безопасность",
        "label_en": "Safety",
        "icon": "shield-check",
    },
    CertificationCategory.ORIGIN: {
        "label_ru": "Происхождение",
        "label_en": "Origin",
        "icon": "map-pin",
    },
    CertificationCategory.OTHER: {
        "label_ru": "Другое",
        "label_en": "Other",
        "icon": "document-text",
    },
}
